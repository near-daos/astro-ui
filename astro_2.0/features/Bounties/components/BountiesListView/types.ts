import { Bounty } from 'types/bounties';
import { ReactNode } from 'react';
import { ProposalVariant } from 'types/proposal';

export type SectionItem = {
  id: string;
  title: string;
  proposer: string;
  proposalId: string;
  bounty?: Bounty;
  link: {
    pathname: string;
    query: {
      dao: string;
      bounty?: string;
      proposal?: string;
    };
  };
  completeHandler?: (
    id: number,
    variant: ProposalVariant.ProposeDoneBounty
  ) => void;
  content: ReactNode;
};
