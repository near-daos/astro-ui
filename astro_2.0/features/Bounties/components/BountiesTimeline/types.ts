export type TimelineMilestoneType =
  | 'Proposal Created'
  | 'Bounty Created'
  | 'Claim'
  | 'Pending Approval'
  | 'Complete Claim'
  | 'Claim Deadline'
  | 'Complete Bounty';

export type TimelineMilestone = {
  type: TimelineMilestoneType;
  date: Date;
  tooltip: string;
  color?: string;
};

export type TimelineClaim = {
  id: string;
  milestones: TimelineMilestone[];
  title: string;
  color: string;
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGroup = {
  id: string;
  name: string;
  isOpen: boolean;
  milestones: TimelineMilestone[];
  claims: TimelineClaim[];
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGranularity = 'month' | 'day' | 'hour';
