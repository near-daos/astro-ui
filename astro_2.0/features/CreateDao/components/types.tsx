import { IconName } from 'components/Icon';

export type DAOType = 'club' | 'cooperative' | 'corporation' | 'foundation';
export type Subject = 'proposals' | 'structure' | 'voting';
export type DAOVotingPowerType = 'democratic' | 'weighted';
export type DAOProposalsType = 'open' | 'closed';
export type DAOStructureType = 'flat' | 'groups';
export type DaoImageType = 'flagCover' | 'flagLogo' | 'tokenImage';

export interface DAOTemplate {
  title: string;
  description: string;
  variant: DAOType;
  proposals: DAOProposalsType;
  structure: DAOStructureType;
  voting: DAOVotingPowerType;
  disabled?: boolean;
}

export type DaoSettingOption<T> = {
  title: string;
  value: T;
  icon: IconName;
  subject: Subject;
  description: string;
  disabled?: boolean;
};

export interface IDaoCreateForm {
  address: string;
  displayName: string;
  purpose: string;
  websites: string[];
  flagCover: FileList;
  flagLogo: FileList;
  defaultFlag: string;
}

export interface DAOFormValues extends IDaoCreateForm {
  proposals: DAOProposalsType;
  structure: DAOStructureType;
  voting: DAOVotingPowerType;
  legalStatus?: string;
  legalLink?: string;
  gas: number;
  members: string[];
}

export type DaoSubjectInfo = {
  subject: Subject;
  title: string;
  subTitle: string;
};
