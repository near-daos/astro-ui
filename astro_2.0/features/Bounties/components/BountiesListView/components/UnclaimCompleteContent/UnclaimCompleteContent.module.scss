.root {
  align-items: center;
  display: flex;
}

.complete {
  margin-left: 12px;
}

.button {
  font-size: 14px;
  text-transform: capitalize;
}
