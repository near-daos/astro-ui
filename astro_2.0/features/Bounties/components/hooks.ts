import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
