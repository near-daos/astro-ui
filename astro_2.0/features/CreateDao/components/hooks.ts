import {
  useFormContext,
  UseFormGetValues,
  UseFormSetValue,
} from 'react-hook-form';
import {
  DAOFormValues,
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
} from 'astro_2.0/features/CreateDao/components/types';
import { useCallback, useEffect, useState } from 'react';
import {
  DAO_PROPOSALS_OPTIONS,
  DAO_STRUCTURE_OPTIONS,
  DAO_VOTING_POWER_OPTIONS,
} from 'astro_2.0/features/CreateDao/components/data';
import { httpService } from 'services/HttpService';
import { CreateDaoCustomInput, CreateDaoInput } from 'types/dao';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import { SputnikWalletError } from 'errors/SputnikWalletError';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import {
  getInitialValues,
  updateAction,
} from 'astro_2.0/features/CreateDao/components/helpers';
import { useStateMachine } from 'little-state-machine';
import { useWalletContext } from 'context/WalletContext';
import { GA_EVENTS, sendGAEvent } from 'utils/ga';

type DaoFormStateReturn = {
  options: DaoSettingOption<
    DAOStructureType | DAOVotingPowerType | DAOProposalsType
  >[];
  getValues: UseFormGetValues<DAOFormValues>;
  setValue: UseFormSetValue<DAOFormValues>;
};

export const useDaoFormState = (): DaoFormStateReturn => {
  const { getValues, watch, setValue } = useFormContext<DAOFormValues>();

  const [structure, voting, proposals] = watch([
    'structure',
    'voting',
    'proposals',
  ]);

  const [options, setOptions] = useState<
    DaoSettingOption<DAOStructureType | DAOVotingPowerType | DAOProposalsType>[]
  >([]);

  useEffect(() => {
    setOptions([
      DAO_STRUCTURE_OPTIONS[structure],
      DAO_VOTING_POWER_OPTIONS[voting],
      DAO_PROPOSALS_OPTIONS[proposals],
    ]);
  }, [structure, voting, proposals, setValue]);

  return { options, getValues, setValue };
};

export function useImageUpload(): {
  uploadImage: (file: File) => Promise<string>;
} {
  const uploadImage = useCallback(async (img: File) => {
    try {
      if (img) {
        const { data } = await httpService.post('/api/upload-to-s3', img, {
          baseURL: '',
        });

        return data;
      }
    } catch (e) {
      return '';
    }

    return '';
  }, []);

  return { uploadImage };
}

export function useCreateDao(): {
  uploadAssets: (defaultFlag: string) => Promise<string>;
  createDao: (
    daoName: string,
    data: CreateDaoInput | CreateDaoCustomInput
  ) => Promise<void>;
} {
  const router = useRouter();
  const { t } = useTranslation();
  const { accountId, nearService } = useWalletContext();
  const { actions, state } = useStateMachine({ updateAction });

  const uploadImg = useCallback(async (img: File) => {
    if (img) {
      const { data } = await httpService.post('/api/upload-to-s3', img, {
        baseURL: '',
      });

      return data;
    }

    return '';
  }, []);

  async function loadImage(defaultFlag: string) {
    const response = await fetch(defaultFlag);
    const blob = await response.blob();

    return new File([blob], 'image.png', { type: 'image/png' });
  }

  const uploadAssets = useCallback(
    async (defaultFlag: string) => {
      try {
        const defaultFlagFile = await loadImage(defaultFlag);

        return uploadImg(defaultFlagFile);
      } catch (e) {
        return '';
      }
    },
    [uploadImg]
  );

  const createDao = useCallback(
    async (daoName: string, data: CreateDaoInput | CreateDaoCustomInput) => {
      try {
        await nearService?.createDao(data);

        const { nearConfig } = configService.get();
        const daoId = `${daoName}.${nearConfig?.contractName ?? ''}`;

        sendGAEvent({
          name: GA_EVENTS.CREATE_DAO,
          daoId,
          accountId,
        });

        showNotification({
          type: NOTIFICATION_TYPES.INFO,
          description: t('notifications.transactionDelay'),
          lifetime: 20000,
        });

        sessionStorage.removeItem('__LSM__');
        actions.updateAction(
          getInitialValues(accountId, state.assets.defaultFlag)
        );

        await router.push(`/dao/${daoId}`);
      } catch (error) {
        console.warn(error);

        if (error instanceof SputnikWalletError) {
          showNotification({
            type: NOTIFICATION_TYPES.ERROR,
            description: error.message,
            lifetime: 20000,
          });
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [router, uploadImg, nearService]
  );

  return {
    createDao,
    uploadAssets,
  };
}
