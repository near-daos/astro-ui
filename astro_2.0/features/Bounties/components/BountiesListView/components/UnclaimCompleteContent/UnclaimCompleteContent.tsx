import React, { FC } from 'react';
import cn from 'classnames';

import { Button } from 'components/button/Button';

import { useBountyControls } from 'astro_2.0/features/Bounties/components/hooks';

import { DAO } from 'types/dao';
import { Bounty } from 'types/bounties';
import { ProposalVariant } from 'types/proposal';

import styles from './UnclaimCompleteContent.module.scss';

interface UnclaimCompleteContentProps {
  dao: DAO;
  bounty: Bounty;
  className?: string;
  completeHandler?: (
    id: number,
    variant: ProposalVariant.ProposeDoneBounty
  ) => void;
}

export const UnclaimCompleteContent: FC<UnclaimCompleteContentProps> = ({
  dao,
  bounty,
  className,
  completeHandler,
}) => {
  const { handleUnclaim } = useBountyControls(dao.id, bounty);

  return (
    <div className={cn(styles.root, className)}>
      <Button
        variant="secondary"
        size="small"
        type="submit"
        data-testid="ucc-unclaim"
        onClick={() => handleUnclaim()}
        className={cn(styles.unclaim, styles.button)}
      >
        Unclaim
      </Button>

      <Button
        variant="black"
        size="small"
        data-testid="ucc-complete"
        onClick={() => {
          if (completeHandler) {
            completeHandler(bounty.bountyId, ProposalVariant.ProposeDoneBounty);
          }
        }}
        className={cn(styles.complete, styles.button)}
      >
        Complete
      </Button>
    </div>
  );
};
