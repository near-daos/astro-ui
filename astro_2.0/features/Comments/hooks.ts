import { useCallback, useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import uniqBy from 'lodash/uniqBy';
import { useAsyncFn, useMount, useMountedState } from 'react-use';
import { useDraftsContext } from 'astro_2.0/features/Drafts/components/DraftsProvider';
import { DraftComment } from 'services/DraftsService/types';
import { useWalletContext } from 'context/WalletContext';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import io, { Socket as TSocket } from 'socket.io-client';
import { fetcher as getDraftProposalComments } from 'services/ApiService/hooks/useDraftComments';
import { useFlags } from 'launchdarkly-react-client-sdk';
import { DraftCommentIndex } from 'services/SearchService/types';
import { mapDraftCommentIndexToDraftComment } from 'services/SearchService/mappers/draftComment';

type Socket = typeof TSocket;

function prepareData(data: DraftComment[]) {
  const groups = data.reduce<Record<string, DraftComment[]>>((res, item) => {
    const key = item.replyTo ?? '';

    if (res[key]) {
      res[key].push(item);
    } else {
      res[key] = [item];
    }

    return res;
  }, {});

  if (!groups['']) {
    return data;
  }

  return groups[''].map(item => {
    return {
      ...item,
      replies: groups[item.id] ?? [],
    };
  });
}

export function useDraftComments(): {
  loading: boolean;
  countComments: number;
  data: DraftComment[];
  addComment: (val: string) => Promise<void>;
  editComment: (val: string, id: string) => Promise<void>;
  deleteComment: (id: string) => Promise<void>;
  likeComment: (id: string, isLiked: boolean) => Promise<void>;
  dislikeComment: (id: string, idDislike: boolean) => Promise<void>;
} {
  const isMounted = useMountedState();
  const router = useRouter();
  const { draft, dao } = router.query;
  const contextId = draft as string;
  const daoId = dao as string;
  const contextType = 'DraftProposal';

  const { useOpenSearchDataApi } = useFlags();
  const { draftsService, setAmountComments } = useDraftsContext();
  const { accountId, pkAndSignature } = useWalletContext();

  const [value, setValue] = useState<{
    data: DraftComment[];
    countComments: number;
  }>({
    data: [],
    countComments: 0,
  });

  const [{ loading }, getAllComments] = useAsyncFn(async () => {
    if (useOpenSearchDataApi === undefined) {
      return;
    }

    try {
      const data = useOpenSearchDataApi
        ? await getDraftProposalComments(
            'draftProposalComment',
            daoId,
            contextId
          )
        : await draftsService.getDraftComments({
            contextId,
            contextType,
            offset: 0,
            limit: 1000,
          });

      setValue({
        data,
        countComments: data.length,
      });
    } catch (e) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: e?.message,
      });
    }
  }, [contextId, useOpenSearchDataApi]);

  useMount(() => getAllComments());

  const addComment = useCallback(
    async (msg: string, replyTo?: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.createDraftComment({
          contextId,
          daoId,
          contextType,
          message: msg,
          replyTo,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const editComment = useCallback(
    async (msg: string, id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.editDraftComment({
          id,
          daoId,
          draftId: contextId,
          message: msg,
          accountId,
          publicKey,
          signature,
        });

        await getAllComments();
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [pkAndSignature, draftsService, daoId, contextId, accountId, getAllComments]
  );

  const likeComment = useCallback(
    async (id: string, isLiked: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isLiked) {
          await draftsService.removeLikeDraftComment(params);
        } else {
          await draftsService.removeDislikeDraftComment(params);
          await draftsService.likeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const dislikeComment = useCallback(
    async (id: string, isDislike: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isDislike) {
          await draftsService.removeDislikeDraftComment(params);
        } else {
          await draftsService.removeLikeDraftComment(params);
          await draftsService.dislikeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const deleteComment = useCallback(
    async (id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.deleteDraftComment({
          id,
          daoId,
          draftId: contextId,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  useEffect(() => {
    setAmountComments(value?.countComments || 0);
  }, [value?.countComments, setAmountComments]);

  useEffect(() => {
    let socket: Socket;
    const { appConfig } = configService.get();

    const { publicKey, signature } = pkAndSignature || {};

    if (accountId && publicKey && isMounted() && appConfig) {
      socket = io(appConfig.DRAFTS_API_URL, {
        query: {
          accountId,
          publicKey,
          signature,
        },
        transports: ['websocket'],
      });

      if (socket) {
        socket.on(
          'draft-comment',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                return {
                  data: uniqBy([newValue, ...prev.data], item => item.id),
                  countComments: prev.countComments + 1,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-updated',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                const newData = prev.data.map(item => {
                  if (item.id !== newValue.id) {
                    return item;
                  }

                  return newValue;
                });

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-removed',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              const newValue = mapDraftCommentIndexToDraftComment(
                comment as DraftCommentIndex
              ) as DraftComment;

              setValue(prev => {
                const newData = prev.data.filter(
                  item => item.id !== newValue.id
                );

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
      }
    }
  }, [accountId, isMounted, pkAndSignature, useOpenSearchDataApi]);

  const preparedData = prepareData(value.data);

  return {
    loading,
    countComments: preparedData?.length || 0,
    data: preparedData ?? [],
    addComment,
    editComment,
    deleteComment,
    likeComment,
    dislikeComment,
  };
}
