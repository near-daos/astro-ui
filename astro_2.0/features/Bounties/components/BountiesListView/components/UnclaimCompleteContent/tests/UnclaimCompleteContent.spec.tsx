/* eslint-disable @typescript-eslint/ban-ts-comment */
import { DAO } from 'types/dao';
import { Bounty } from 'types/bounties';
import { ProposalVariant } from 'types/proposal';

import { render } from 'jest/testUtils';

import { useBountyControls } from 'astro_2.0/features/Bounties/components/hooks';

import { UnclaimCompleteContent } from 'astro_2.0/features/Bounties/components/BountiesListView/components/UnclaimCompleteContent';
import { fireEvent } from '@testing-library/dom';

jest.mock('astro_2.0/features/Bounties/components/hooks', () => {
  return {
    useBountyControls: jest.fn(() => ({})),
  };
});

describe('UnclaimCompleteContent', () => {
  const bountyId = 'bountyId';

  const dao = {} as unknown as DAO;
  const bounty = {
    bountyId,
  } as unknown as Bounty;

  it('Should render component', () => {
    const { container } = render(
      <UnclaimCompleteContent dao={dao} bounty={bounty} />
    );

    expect(container).toMatchSnapshot();
  });

  it('Should call unclaim callback', () => {
    const handleUnclaim = jest.fn();

    // @ts-ignore
    useBountyControls.mockImplementation(() => ({
      handleUnclaim,
    }));

    const { getByTestId } = render(
      <UnclaimCompleteContent dao={dao} bounty={bounty} />
    );

    fireEvent.click(getByTestId('ucc-unclaim'));

    expect(handleUnclaim).toBeCalled();
  });

  it('Should call complete callback', () => {
    const completeHandler = jest.fn();

    const { getByTestId } = render(
      <UnclaimCompleteContent
        dao={dao}
        bounty={bounty}
        completeHandler={completeHandler}
      />
    );

    fireEvent.click(getByTestId('ucc-complete'));

    expect(completeHandler).toBeCalledWith(
      bountyId,
      ProposalVariant.ProposeDoneBounty
    );
  });

  it('Should not fail if no complete callback', () => {
    const { getByTestId } = render(
      <UnclaimCompleteContent dao={dao} bounty={bounty} />
    );

    fireEvent.click(getByTestId('ucc-complete'));
  });
});
