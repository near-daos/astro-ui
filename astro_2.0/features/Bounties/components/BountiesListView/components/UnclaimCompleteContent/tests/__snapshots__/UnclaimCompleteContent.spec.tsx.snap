// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnclaimCompleteContent Should render component 1`] = `
<div>
  <div
    class="root"
  >
    <button
      class="btn secondary sizeSmall unclaim button"
      data-testid="ucc-unclaim"
      type="submit"
    >
      Unclaim
    </button>
    <button
      class="btn black sizeSmall complete button"
      data-testid="ucc-complete"
      type="button"
    >
      Complete
    </button>
  </div>
</div>
`;
